### PR TITLE
Feature/li siodist

### LINKED ISSUE #1039: Add a capability to write non-synchronized distributed output in LIS
The I/O is a significant bottleneck in the LIS runs, particularly for model runs over large spatial extents and/or fine resolutions. This is partly because of the synchronization inherent within the code to assemble variables across multiple processors before writing them to an external file. The proposed capability would eliminate this process, where each processor would write out the data for the local domain. The quilting of the data would then be conducted by an external program (such as LVT).

### PR DESCRIPTION
**Pull request by sujayvkumar**
### Description

A new output format called 'distributed binary' was added. In this format, each processor writes outputs for the respective local domain to a binary file. At the output time steps, no synchronization is done to gather output variables across multiple processors. 

The quilting of the outputs across multiple processors is then done using LVT using a new run mode called "LIS post processing"

Resolves #1039 

### Testcase

A testcase can be found at : /discover/nobackup/projects/eis_freshwater/svkumar/Tests/LIS_test_distio

the lisrun directory contains the LIS example config. In this setup 16 processors should be used. 
the lvtrun directory then contains the configuration to quilt the files into a single file in netcdf format. 

The outputs will be generated in the same directory as that of the LIS output.

### CODE REVIEW

File: lis/core/LIS_fileIOMod.F90 [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ! Land Information System Framework (LISF)
 ! Version 7.3
 !
-! Copyright (c) 2020 United States Government as represented by the
+! Copyright (c) 2015 United States Government as represented by the
```
Comment (bmcandr) [inline lis/core/LIS_fileIOMod.F90:6]: Update copyright

File: lis/core/LIS_fileIOMod.F90 [hunk 2]
```diff
@@ -415,9 +415,21 @@ subroutine create_output_filename(n, fname, model_name, odir, writeint)
             out_fname = trim(out_fname)//'.ts4r'
          elseif(LIS_rc%wopt.eq."2d gridspace") then 
             out_fname = trim(out_fname)//'.gs4r'
+         elseif(LIS_rc%wopt.eq."2d ensemble gridspace") then 
+            out_fname = trim(out_fname)//'.gs4r'
          elseif(LIS_rc%wopt.eq."1d gridspace") then 
             out_fname = trim(out_fname)//'.gs4r'
          endif
+      case ("distributed binary")
```
Comment (bmcandr) [inline lis/core/LIS_fileIOMod.F90:423]: Add new output data format to docs

File: lis/core/LIS_histDataMod.F90 [hunk 3]
```diff
@@ -2073,6 +2075,19 @@ subroutine LIS_histDataInit(n, ntiles)
             model_patch=.true.)
     endif
 
+    call ESMF_ConfigFindLabel(modelSpecConfig,"VPD:",rc=rc)
+    call get_moc_attributes(modelSpecConfig, LIS_histData(n)%head_lsm_list, &
+         "VPD",&
+         "vapor_pressure_deficit",&
+         "vapor pressure deficit",rc)
```
Comment (bmcandr) [inline lis/core/LIS_histDataMod.F90:2082]: Support for VPD output seems unrelated to distributed binary support. No need to change, just adding a comment so it is acknowledged in the PR.

File: lis/core/LIS_historyMod.F90 [hunk 4]
```diff
@@ -451,10 +451,11 @@ subroutine LIS_writeModelOutput(n, lsmoutfile, lsmstatsfile, &
 !   \end{description}
 !EOP
 
-    integer :: ftn, ftn_stats
+    integer :: ftn, ftn_stats, ftnp(LIS_npes)
     integer :: iret
     integer :: group_temp
-    character*100 :: mname_temp
+    character*100 :: mname_temp,temp1,lsmoutfilep
```
Comment (bmcandr) [inline lis/core/LIS_historyMod.F90:457]: Declare path/filename variables with length = `LIS_CONST_PATH_LEN`

File: lvt/core/LVT_LISpostMod.F90 [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+
+    call ESMF_ConfigGetAttribute(LVT_config,LVT_LISpost%npes,&
+         label="Number of processors used in the LIS output generation:",rc=rc)
+    call LVT_verify(rc,'Number of processors used in the LIS output generation: option not specified in the config file')
+    call ESMF_ConfigGetAttribute(LVT_config,LVT_LISpost%nfields,&
+         label="Number of fields in the LIS output:",rc=rc)
+    call LVT_verify(rc,'Number of fields in the LIS output: option not specified in the config file')
+    call ESMF_ConfigGetAttribute(LVT_config,LVT_LISpost%nlevels,&
+         label="Number of vertical levels in the LIS output:",rc=rc)
+    call LVT_verify(rc,'Number of vertical levels in the LIS output: option not specified in the config file')
```
Comment (bmcandr) [inline lvt/core/LVT_LISpostMod.F90:100]: Add new config entries to docs

File: lvt/core/LVT_LISpostMod.F90 [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+          write(temp1,'(i4.4)') i-1
+          read(temp1,fmt='(4a1)') fproc
+          
+          fname(i) = trim(froot)//'.'//fproc(1)//fproc(2)//fproc(3)//fproc(4)
+          
+          inquire(file=trim(fname(i)),exist=file_exists)
+       
+          if(file_exists) then 
+             
+!             write(LVT_logunit,*) '[INFO] Reading LIS output ',trim(fname(i))
```
Comment (bmcandr) [inline lvt/core/LVT_LISpostMod.F90:199]: Might be good to leave this log message in. Is there a reason for commenting it out?

File: lvt/core/LVT_LISpostMod.F90 [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+               nf90_float,dimids=dimID(1),varid=lonid),&
+               'nf90_def_att for lat failed in LVT_LISpostMod')    
+          
+    ! Global attributes
+!    call LVT_verify(nf90_put_att(ftn_nc,NF90_GLOBAL,"NUM_SOIL_LAYERS", &
+!         nsoillayers),&
+!         'nf90_put_att for title failed in LVT_LISpostMod')
+!    call LVT_verify(nf90_put_att(ftn_nc,NF90_GLOBAL,"SOIL_LAYER_THICKNESSES", &
+!         lyrthk),&
+!         'nf90_put_att for title failed in LVT_LISpostMod')
```
Comment (bmcandr) [inline lvt/core/LVT_LISpostMod.F90:274]: Does adding these global attributes not work?

File: lvt/core/LVT_fileIOMod.F90 [hunk 8]
```diff
@@ -356,6 +356,18 @@ subroutine create_output_filename(n, source, fname, model_name, writeint, &
                  out_fname = trim(out_fname)//'.ts4r'
               elseif(LVT_LIS_rc(source)%wopt.eq."2d gridspace") then 
                  out_fname = trim(out_fname)//'.gs4r'
+              elseif(LVT_LIS_rc(source)%wopt.eq."2d ensemble gridspace") then
```
Comment (bmcandr) [inline lvt/core/LVT_fileIOMod.F90:359]: Add new output methodology (`2d ensemble gridspace`) and format (`distributed binary`) to docs

File: lvt/core/LVT_histDataMod.F90 [hunk 9]
```diff
@@ -1009,6 +1009,7 @@ subroutine LVT_histDataInit()
 
     call ESMF_ConfigGetAttribute(LVT_config,LVT_rc%sp_avg_mode, &
          label="Spatial averaging mode:",&
+         default="pixel-by-pixel",&
```
Comment (bmcandr) [inline lvt/core/LVT_histDataMod.F90:1012]: Note new default value in docs

File: lvt/core/LVT_readConfig.F90 [hunk 10]
```diff
@@ -367,6 +367,7 @@ subroutine LVT_readConfig(configfile)
   
   call ESMF_ConfigGetAttribute(LVT_config,LVT_rc%smoothObs, &
        label="Apply temporal smoothing to obs:",&
+       default = 0,&
```
Comment (bmcandr) [inline lvt/core/LVT_readConfig.F90:370]: Note new default values in docs

**Review by bmcandr — CHANGES_REQUESTED** [review]

**Comment by bmcandr** [comment]
Hi @sujayvkumar, I have completed my code review. I identified a few changes that I can take care of (mostly docs related). Two questions for you:

* In `lvt/core/LVT_LISpostMod.F90`:
    *  A log message is commented out. Is there a reason it is commented out? It might be worth keeping unless it spams the log.
    * The calls writing the global attributes `NUM_SOIL_LAYERS`, `SOIL_LAYER_THICKNESSES`, and `history` are disabled. Do they not work? (or are unnecessary?)

**Comment by sujayvkumar** [comment]
@bmcandr : Thanks for the quick review. 
The log is commented out because it could lead to a ton of messages, particularly if the LIS run was done using a large number of processors. 

Yes - they are turned off because they are not present in the distributed binary outputs (and we don't want to include it there since this information doesn't change between outputs). I suppose we can make them configurable options in LVT and then embed them within the netcdf file.

File: lvt/core/LVT_LISpostMod.F90 (hunk 7 above)
Comment (bmcandr) [reply lvt/core/LVT_LISpostMod.F90:274]: See comment below. Will be fixed later.

**Review by bmcandr — APPROVED** [review]

**Comment by bmcandr** [comment]
I am able to reproduce outputs with the Intel 21 compiler. Merging.

I will address the global attributes issue mentioned above in a subsequent PR.

**Pull request merged**